presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchange.java

```diff
@@ -0,0 +1,584 @@
+            long bytes = bufferBytes.addAndGet(delta);
+            if (delta < 0 && bytes < maxBufferedBytes) {
+                writeWaiter.notifyWaiters();
+            }
+        };
+
+        List<Consumer<PageReference>> buffers = this.sources.stream()
+                .map(buffer -> (Consumer<PageReference>) buffer::addPage)
+                .collect(toImmutableList());
+
+        if (partitioning.equals(SINGLE_DISTRIBUTION)) {
+            exchangerSupplier = () -> new BroadcastExchanger(buffers, memoryTracker);
+        }
+        else if (partitioning.equals(FIXED_BROADCAST_DISTRIBUTION)) {
+            exchangerSupplier = () -> new BroadcastExchanger(buffers, memoryTracker);
+        }
+        else if (partitioning.equals(FIXED_RANDOM_DISTRIBUTION)) {
+            exchangerSupplier = () -> new RandomExchanger(buffers, memoryTracker);
+        }
+        else if (partitioning.equals(FIXED_HASH_DISTRIBUTION)) {
+            exchangerSupplier = () -> new PartitioningExchanger(buffers, memoryTracker, types, partitionChannels, partitionHashChannel);
+        }
+        else {
+            throw new IllegalArgumentException("Unsupported local exchange partitioning " + partitioning);
+        }
+    }
+
+    public List<Type> getTypes()
+    {
+        return types;
+    }
+
+    public int getBufferCount()
+    {
+        return sources.size();
+    }
+
+    public long getBuferedBytes()
+    {
+        return bufferBytes.get();
+    }
+
+    public synchronized LocalExchangeSinkFactory createSinkFactory()
+    {
+        checkState(!noMoreSinkFactories, "No more sink factories already set");
+        LocalExchangeSinkFactory sinkFactory = new LocalExchangeSinkFactory(this);
+        sinkFactories.add(sinkFactory);
+        return sinkFactory;
+    }
+
+    public LocalExchangeSource getSource(int partitionIndex)
+    {
+        return sources.get(partitionIndex);
+    }
+
+    private void sourceFinished()
+    {
+        checkNotHoldsLock(this);
+
+        if (!sources.stream().allMatch(LocalExchangeSource::isFinished)) {
+            return;
+        }
+
+        // all sources are finished, so finish the sinks
+        ImmutableList<LocalExchangeSink> openSinks;
+        synchronized (this) {
+            allSourcesFinished = true;
+
+            // since all sources are finished there is no reason to all new pages
+            openSinks = ImmutableList.copyOf(sinks);
+            sinks.clear();
+        }
+
+        openSinks.forEach(LocalExchangeSink::finished);
+        checkAllSinksComplete();
+    }
+
+    private LocalExchangeSink createSink(LocalExchangeSinkFactory factory)
+    {
+        checkNotHoldsLock(this);
+
+        // Note: exchanger can be stateful so create a new one for each sink
+        LocalExchangeSink sink = new LocalExchangeSink(this, exchangerSupplier.get());
+        synchronized (this) {
+            checkState(sinkFactories.contains(factory), "Factory is already closed");
```

Comment:
Compared to checking `factory.closed`, checking `sinkFactories.contains` is
1) more expensive
2) catches less errors (update to `sinkFactories` can be potentially delayed because call to `exchange.sinkFactoryClosed(this)` is done in a separate sync block)